google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -65,11 +83,189 @@ def __init__(self, **kwargs):
+            ('sourceRanges', self._source_ranges),
+            ('sourceTags', self._source_tags),
+            ('targetTags', self._target_tags),
+            ('destinationRanges', self._destination_ranges)
+        ]
+        for key, value in max_256_value_keys:
+            if value and len(value) > 256:
+                raise InvalidFirewallRuleError(
+                    'Rule entry "%s" must contain 256 or fewer values: "%s".'
+                    % (key, value))
+
+        if self._source_tags:
+            if self._source_service_accounts or self._target_service_accounts:
+                raise InvalidFirewallRuleError(
+                    'sourceTags cannot be set when source/targetServiceAccounts'
+                    ' are set')
+
+        if self._target_tags:
+            if self._source_service_accounts or self._target_service_accounts:
+                raise InvalidFirewallRuleError(
+                    'sourceTags cannot be set when source/targetServiceAccounts'
+                    ' are set')
+
+        max_1_value_keys = [
+            ('sourceServiceAccount', self.source_service_accounts),
+            ('targetServiceAccount', self.target_service_accounts),
+        ]
+        for key, value in max_1_value_keys:
+            if value and len(value) > 1:
+                raise InvalidFirewallRuleError(
+                    'Rule entry "%s" may contain at most 1 value: "%s".'
+                    % (key, value))
+
+    def _validate_direction(self):
+        """Checks that the direction and associated fields are valid.
+
+        Raises:
+          InvalidFirewallRuleError: If:
+            * Direction is 'ingress' and
+              * there are no source ranges or tags
+              * _destination_ranges is not set
+            * Direction is 'egress' and
+              * there are no source ranges or tags
+              * _destination_ranges is set
+        """
+        if self.direction == 'ingress':
+            if (not self._source_ranges and not self._source_tags and not
+                    self.source_service_accounts):
+                raise InvalidFirewallRuleError(
+                    'Ingress rule missing required field oneof '
+                    '"sourceRanges" or "sourceTags": "%s".' % self)
+
+            if self._destination_ranges:
+                raise InvalidFirewallRuleError(
+                    'Ingress rules cannot include "destinationRanges": "%s".'
+                    % self)
+
+        elif self.direction == 'egress':
+            if not self._destination_ranges:
+                raise InvalidFirewallRuleError(
+                    'Egress rule missing required field "destinationRanges":'
+                    '"%s".'% self)
+
+            if (self._source_ranges or self._source_tags or
+                    self._source_service_accounts):
+                raise InvalidFirewallRuleError(
+                    'Egress rules cannot include "sourceRanges", "sourceTags":'
```

Comment:
Update message to include sourceServiceAccounts

Reply:
Done.